samza-core/src/main/java/org/apache/samza/container/RunLoopTask.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.container;
+
+import java.util.Collections;
+import org.apache.samza.checkpoint.OffsetManager;
+import org.apache.samza.scheduler.EpochTimeScheduler;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.task.ReadableCoordinator;
+import org.apache.samza.task.TaskCallbackFactory;
+import scala.collection.JavaConversions;
+
+
+public interface RunLoopTask {
```

Comment:
Can we add some java docs for this class?

Reply:
yes, i'll add add a description to the class declaration.

Reply:
Let us take this opportunity to document more details about runloop and its interaction with the implementations of this interface.

1. Exclusivity between process, window, scheduler, commit. & endOfStream with the exception of async commit.
2. Not thread safe and needs synchronization of shared objects between instances, between commit & other methods in case of async commit enabled within an instance
3. Lifecycle of this class if applicable; which brings a question on does this need `init()` and `close()`;

Reply:
+1 regarding adding docs to each method. Ideally, there would have already been more docs in `TaskInstance`, but it is more useful now that this is an interface.

Reply:
I've updated with better class-level and method-level docs.

Reply:
Regarding life cycle... RunLoop currently does not manage life cycle of the tasks it executes. This is done at the scope of SamzaContainer (i.e. the entity creating and running the RunLoop).

Since this interface is targeted to the relationship between RunLoop and the tasks it executes, I don't think life cycle management belongs here.

What do you think @mynameborat @cameronlee314 ?

Reply:
Discussed with @mynameborat and we think that unifying on lifecycle management will (eventually) be more necessary once we've moved side inputs back to using run loop (though a different instance at first). To consolidate to only a single RunLoop instance within container, we'd want some unified way to init/close a RunLoopTask whether it be a TaskInstance or "NotYetImplementedSideInputTask".

@cameronlee314 Feel free to offer your thoughts but for now I think we can consider the life cycle issue out of scope for this PR.